ai_services/anomaly_detection/data_preprocessing_examples/oci_data_flow_based_examples/example_code/ad_utils.py

```diff
@@ -0,0 +1,116 @@
+    token_path = sc.sparkContext.getConf().get(token_key)
+    return token_path
+
+
+def get_authenticated_client(token_path, client, file_location=DEFAULT_LOCATION, profile_name=DEFAULT_PROFILE,
+                             **kwargs):
+    if token_path is None:
+        # You are running locally, so use our API Key.
+        config = oci.config.from_file(file_location, profile_name)
+        kwargs['config'] = config
+        authenticated_client = client(**kwargs)
+    else:
+        # You are running in Data Flow, so use our Delegation Token.
+        with open(token_path) as fd:
+            delegation_token = fd.read()
+        signer = oci.auth.signers.InstancePrincipalsDelegationTokenSigner(
+            delegation_token=delegation_token
+        )
+        kwargs['config'] = {}
+        kwargs['signer'] = signer
+        authenticated_client = client(**kwargs)
+    return authenticated_client
+
+
+class AdUtils:
+    def __init__(self, profile_name=DEFAULT_PROFILE, service_endpoint=None):
+        token_path = get_token_path()
+
+        client_args = {'profile_name': profile_name}
+        self.obj_client = get_authenticated_client(token_path, ObjectStorageClient, **client_args)
+
+        if service_endpoint:
+            client_args['service_endpoint'] = service_endpoint
+        self.ad_client = get_authenticated_client(token_path, AnomalyDetectionClient, **client_args)
+
+    def train(self, project_id, compartment_id, data_assets, target_fap=DEFAULT_TARGET_FAP,
+              training_fraction=DEFAULT_TRAINING_FRACTION):
+        assert data_assets['type'] == 'object_storage', "Unknown dataset details"
+        list_objects_response = self.obj_client.list_objects(namespace_name=data_assets['namespace'],
+                                                             bucket_name=data_assets['bucket'],
+                                                             prefix=data_assets['prefix'])
+        assert list_objects_response.status == 200, f'Error listing objects: {list_objects_response.text}'
+        objects_details = list_objects_response.data
+
+        model_ids = []
+        for object_details in objects_details.objects:
+            if object_details.name.endswith('.csv'):
+                data_asset_id = self._create_data_asset_(project_id=project_id, compartment_id=compartment_id,
+                                                         namespace=data_assets['namespace'],
+                                                         bucket=data_assets['bucket'],
+                                                         object_name=object_details.name)
+                model_id = self._create_model_(project_id=project_id, compartment_id=compartment_id,
+                                               data_asset_id=data_asset_id, target_fap=target_fap,
+                                               training_fraction=training_fraction)
+                model_ids.append(model_id)
```

Comment:
Actually, the control won't go to line 79 in that case. It will cause an [`AssertionError`](https://github.com/bug-catcher/oci-data-science-ai-samples/blob/863c43d6f0fcaa86e21a013d8338c5eda0ca1ba2/ai_services/anomaly_detection/data_preprocessing_examples/oci_data_flow_based_examples/example_code/ad_utils.py#L63-L64) and the control will flow directly to [`except`](https://github.com/bug-catcher/oci-data-science-ai-samples/blob/2e0951c3086632780ba0fc4065f5d54095cebf1b/ai_services/anomaly_detection/data_preprocessing_examples/oci_data_flow_based_examples/example_code/df_driver.py#L193-L194) clause.